### PR TITLE
global: allow skip single elements

### PR DESCRIPTION
**Pull request by david-caro**
- Add the possibility to skip single elements when using the for_each
  decorator, either raising the IgnoreElem exception or using the
  ignore_list_elem decorator.

Signed-off-by: David Caro david@dcaro.es

### CODE REVIEW

File: dojson/errors.py [hunk 1]
```diff
@@ -26,6 +26,14 @@ class IgnoreKey(DoJSONException):
     """
 
 
+class IgnoreElem(DoJSONException):
+    """The corresponding elem for the current list key has been ignored.
+
+    .. versionadded:: ??
```
Comment (david-caro) [inline dojson/errors.py]: I'll need a little help here to figure out what to put in

Comment (jirikuncar) [inline dojson/errors.py]: next minor version, hence `1.3.0`

**Comment by jirikuncar** [comment]
- [x] Needs tests.
- [x] `s/Add the possibility/Adds the possibility`
- [x] Add yourself to https://github.com/inveniosoftware/dojson/blob/master/AUTHORS.rst
- [ ] This branch has conflicts that must be resolved.

File: dojson/errors.py [hunk 2]
```diff
@@ -26,6 +26,14 @@ class IgnoreKey(DoJSONException):
     """
 
 
+class IgnoreElem(DoJSONException):
+    """The corresponding elem for the current list key has been ignored.
```
Comment (jirikuncar) [inline dojson/errors.py]: elem -> element

File: dojson/errors.py [hunk 3]
```diff
@@ -26,6 +26,14 @@ class IgnoreKey(DoJSONException):
     """
 
 
+class IgnoreElem(DoJSONException):
```
Comment (jirikuncar) [inline dojson/errors.py]: What about calling it: `IgnoreItem`

File: dojson/utils.py [hunk 4]
```diff
@@ -44,6 +44,21 @@ def wrapper(self, key, value, **kwargs):
     return wrapper
 
 
+def ignore_list_elem(f):
```
Comment (jirikuncar) [inline dojson/utils.py]: IMHO this decorator is not needed. You can directly raise in the function implementing a rule.

File: dojson/errors.py [hunk 5]
```diff
@@ -26,6 +26,14 @@ class IgnoreKey(DoJSONException):
...
+    """The corresponding elem for the current list key has been ignored.
+
+    .. versionadded:: ??
+
```
Comment (jirikuncar) [inline dojson/errors.py:29]: Is the empty line really needed?

File: dojson/errors.py (hunk 3 above)
Comment (david-caro) [inline dojson/errors.py]: To me they look the same to me (not better, not worse).

File: dojson/errors.py (hunk 5 above)
Comment (david-caro) [inline dojson/errors.py:29]: no, it's really needed for it not to be there?

File: dojson/utils.py (hunk 4 above)
Comment (david-caro) [inline dojson/utils.py]: I just followed the current code structure, to skip None return values you have the ignore_value above. IMO it's easier to use, as you don't have to know the implementation behind it (using exceptions), you can just rely on this decorator to do whatever is needed.

File: dojson/errors.py (hunk 1 above)
Comment (david-caro) [inline dojson/errors.py]: :+1:

File: dojson/errors.py (hunk 3 above)
Comment (jirikuncar) [inline dojson/errors.py]: then we should call it `element` instead of `elem` however I still prefer `item`

File: dojson/utils.py (hunk 4 above)
Comment (jirikuncar) [inline dojson/utils.py]: IMHO the usage is the same as raising `IgnoreKey` exception.

File: dojson/errors.py (hunk 5 above)
Comment (jirikuncar) [inline dojson/errors.py:29]: Good. Then please remove it so we don't have unnecessary empty lines in the code that are not helping with readability.

Comment (david-caro) [inline dojson/errors.py:29]: Should I remove from all the other comments that also have it? In this same pr?

Comment (jirikuncar) [inline dojson/errors.py:29]: After you remove `ignore_list_elem` there is no other place.

Comment (david-caro) [inline dojson/errors.py:29]: In this same file, lines 17,  25, 41, and in utils.py line 36

Comment (david-caro) [reply dojson/errors.py:29]: Yes, there is, lines 17, 25 and 41 of this file, and 36 of utils.py

File: tests/test_core.py [hunk 6]
```diff
@@ -289,6 +294,25 @@ def match(self, key, value):
...
+def test_none_element():
+    """Test none value for a list element."""
+    overdo = dojson.Overdo()
+
+    @overdo.over('b', 'b')
+    @for_each_value
+    @ignore_list_element
+    def match(self, key, value):
+        if value == 2:
+            return None
```
Comment (jirikuncar) [inline tests/test_core.py:306]: Please raise `IgnoreElement` rather than returning `None`.

Comment (david-caro) [reply tests/test_core.py:306]: I'm using the `ignore_list_element` decorator, that captures the None values, just as the `ignore_value` one does.

File: dojson/utils.py [hunk 7]
```diff
@@ -44,6 +43,20 @@ def wrapper(self, key, value, **kwargs):
     return wrapper
 
 
+def ignore_list_element(f):
```
Comment (jirikuncar) [inline dojson/utils.py:46]: I'm still not convinced that we need this decorator. Please remove it from this PR and we can start separate discussion on its usage. Thanks

File: dojson/utils.py [hunk 8]
```diff
@@ -44,6 +43,20 @@ def wrapper(self, key, value, **kwargs):
...
+    """Skip empty element for the current key for None value.
+
+    .. versionadded:: 0.3.0
```
Comment (jirikuncar) [inline dojson/utils.py:49]: (1.3.0)

File: tests/test_core.py (hunk 6 above)
Comment (jirikuncar) [reply tests/test_core.py:306]: see https://github.com/inveniosoftware/dojson/pull/169#discussion_r75817846

File: tests/test_core.py [hunk 9]
```diff
@@ -20,7 +20,12 @@
 from dojson.contrib.marc21.utils import create_record, load, split_stream
 from dojson.contrib.to_marc21 import to_marc21
 from dojson.contrib.to_marc21.utils import dumps
-from dojson.utils import flatten, for_each_value, ignore_value
+from dojson.utils import (
```
Comment (jirikuncar) [inline tests/test_core.py:23]: please use `isort` with settings in `.editorconfig`

**Comment by jirikuncar** [comment]
closed via #175

**Pull request closed without merging**